### PR TITLE
Custom labels docs

### PR DESCRIPTION
**Pull request by UncleGedd**
### Decorator Syntax Examples
- Adds documentation and working example around the functions mentioned in the comments of this [issue](https://github.com/apache/incubator-flagon-useralejs/issues/63)
- IMO the second method is simpler than than the first, would like to get some user feedback to see if this is adds value with respect to custom label implementation

### CODE REVIEW

File: example/log-label-example/README.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+```js
+document.addEventListener('click', function (e) {
+    if (e.target.innerHTML === 'New Feature') {
+        window.userale.map(log => ({...log, logType: 'custom', customLabel: 'New Feature'}));
+        window.userale.packageLog(e, window.userale.details(window.userale.options(), e.type));
+        window.userale.map();
+    } else {
+        return false
+    }
+});
```
Comment (confusingstraw) [inline example/log-label-example/README.md:29]: this works, but i think it relies a bit on the implementation details of `map`/`packageLog`. the intended usage is captured in the (non-highlighted) example in the main `README.md`.

when we call the `userale.map` function, it keeps the function you pass it around until it gets re-assigned. the given function gets called for _every_ log that the library receives. with that in mind, there is no need to setup your own event listener, as that is what userale is for. all you need to 

here is how i'd imagine the above code would be implemented:
```js
// no need to set up a separate listener
window.userale.map((log, e) => {
    // determine whether we want to modify the log
    if (e.target.innerHTML !== 'New Feature') {
        return log; // return the log if we don't care about this event
    }

    // return the log with any changes we want
    return {
        ...log,
        customLabel: 'New Feature',
        logType: 'custom',
    };
});
```

Comment (confusingstraw) [reply example/log-label-example/README.md:29]: if you want to make the code extra compact (not something i am a huge fan of), you can write it like this:
```js
window.userale.map((log, e) => e.target.innerHTML === 'New Feature'
    ? { ...log, customLabel: 'New Feature', logType: 'custom' }
    : log);
```

Comment (UncleGedd) [reply example/log-label-example/README.md:29]: Agreed Rob, I also prefer the idea of just using the `userale.map` function as opposed to adding more event listeners. I think we need a bit more context on what exactly the users are trying to accomplish and what they want to [decorate](https://github.com/apache/incubator-flagon-useralejs/issues/63).

The examples that Rob linked and are located at [example/index.js](https://github.com/apache/incubator-flagon-useralejs/blob/master/example/index.js). Do you have any idea why in that code the extra event listeners are added?

**Review by confusingstraw — COMMENTED** [review]
>I saw the examples you mentioned over in the README but I'm not quite understanding their usage. Is the goal just to subvert the Userale logs and event handlers?
the goal is to hook into the existing userale logging pipeline. `userale.filter` is if you want to exclude certain logs from being sent.
for example, let's say you didn't want _any_ click events, you might use `filter`:
```js
window.userale.filter(log => log.type !== 'click')
```
or if you have some extra application-specific data you want to include with the logs:
```js
window.userale.map(log => { ...log, isAdmin: myApplicationState.isAdmin });
```
and you may potentially want to tie them together!
```js
window.userale.filter(log => log.type !== 'click');
window.userale.map(log => { ...log, isAdmin: myApplicationState.isAdmin });
```
^ this will result in only capturing events that are _not_ clicks, and will add an `isAdmin` property to all the generated logs.

the reason it is setup as a one-off function that gets assigned is that, when i designed it, i didn't want developers to have to explicitly write an event listener/manually package a log just to augment our existing logs. otherwise, why would they use userale at all if they still have to set all of their own listeners?

let me know if you have any other questions.

**Comment by UncleGedd** [comment]
Gotcha, thanks for the thorough explanation! Understood that the Userale map/filter functions operate on the logging pipeline and they operate just like their JS array API counterparts. In the example code above, what is the purpose of adding the event listener?

**Comment by poorejc** [comment]
I built those examples just to demonstrate some simple custom logging behavior for .log, .packageLog, .packageCustomLog exports. Also wanted some simple examples for how UserAle performed against form data, which wasn’t previously well supported. Two birds with one stone.

But, yes you’re not crazy—those event handlers are duplicative of out of the box UserAle event capture. Just super simple examples that are easy to understand in a very simple webpage. Clearly, maybe not as simple as I hoped ;)

Joshua Poore

**Comment by confusingstraw** [comment]
> Gotcha, thanks for the thorough explanation! Understood that the Userale map/filter functions operate on the logging pipeline and they operate just like their JS array API counterparts. In the example code above, what is the purpose of adding the event listener? (noting that the snippet was originally taken from [examples/index.js](https://github.com/apache/incubator-flagon-useralejs/blob/master/example/index.js))

as @poorejc mentioned, it is probably just an artifact of quickly writing examples. the only examples i'm familiar with are the first two on the main `README.md` examples section. it is my fault for not having made a larger body of examples to pull from, as the `map`/`filter`/`packageCustomLog` API should be flexible enough to handle the vast majority of use cases, and it should be possible to do so without writing hacks.

**Comment by poorejc** [comment]
I will note that I have observed some users actually build their own event handlers and fully customize their logs—I didn’t advocate this, but that’s why I modified packageLog and built packageCustomLog—to help enforce data schema consistency even among users who insist on a “tag manager” sort of feel. Not intended use, but not insane.

I do agree that those examples might make it confusing re what UserALE does natively.

I’ll pass on another example tonight I built for handling an event that we actually don’t listen for (and probably never will add to ./src/attachHandlers.

Joshua Poore

File: example/log-label-example/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# Custom Log Label Examples
+
+Examples illustrating how to add custom labels to logs generated by `userale`.
```
Comment (poorejc) [inline example/log-label-example/README.md:3]: should be consistent in naming `userale` --> `useralejs` OR `UserALE.js`

Comment (poorejc) [inline example/log-label-example/README.md:3]: I would also add a little more context to the usage. 

```
The following are examples illustrating how to add custom labels to log generated by `UserALE.js`. Custom labels are useful in cases where one would prefer to use a more descriptive label for an element users interact with than HTML tags (among other cases).
```

File: example/log-label-example/README.md (hunk 1 above)
Comment (poorejc) [reply example/log-label-example/README.md:29]: +1  For labeling, we should not advocate that users create new event listeners. Ideally, users should only be creating new events listeners if (1) the events are not captured by UserALE.js natively, or (2) they want to filter out everything but custom logs.

File: example/log-label-example/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    <button>New Feature</button>
+</div>
+```
+
+The following code snippet will add a custom field and send a log whenever the new feature button is clicked:
+
+```js
+document.addEventListener('click', function (e) {
+    if (e.target.innerHTML === 'New Feature') {
+        window.userale.map(log => ({...log, logType: 'custom', customLabel: 'New Feature'}));
```
Comment (poorejc) [inline example/log-label-example/README.md:23]: window.userale.map is fine for now. Eventually, I think we'll want to encourage (or promote) more usage of our NPM deployment option (vs script tag). If including UserALE.js via NPM package, exports do not have to be called as a window object, e.g., window.userale.map --> userale.map. I think this is fine for now as it's consistent usage with our "top-level" example (which highlights script-tag use case).

File: example/log-label-example/README.md (hunk 1 above)
Comment (poorejc) [reply example/log-label-example/README.md:29]: > if you want to make the code extra compact (not something i am a huge fan of), you can write it like this:
> 
> ```js
> window.userale.map((log, e) => e.type ==='click' && e.target.innerHTML === 'New Feature'
>     ? { ...log, customLabel: 'New Feature', logType: 'custom' }
>     : log);
> ```

+1 agree with Rob. Worth being verbose in examples for accessibility and descriptiveness. If they want to compact our examples, that's not hard.

**Review by poorejc — COMMENTED** [review]
Looks good to me. Rob's comments are spot on. Added minor mods in comments on that. If you want to make those changes and push another commit, easy to approve and I'm happy to merge to test.

**Comment by UncleGedd** [comment]
All, thanks for the feedback! I will modify the documentation removing the event listener parts and clean up the verbiage to be more consistent.

**Comment by UncleGedd** [comment]
Hi all, while I was implementing the proposed changes I came across a curious behavior. Currently, `packageCustomLog` does not take the event as an argument and, as a result,  does not pass the event to the `mapHandler` function:

https://github.com/apache/incubator-flagon-useralejs/blob/7a7b2ce73963d9dd1b0bff9419ded12e15af10ad/src/packageLogs.js#L165-L166

However, in `packageLog` we do take the event as an argument and pass it to the `mapHandler`:

https://github.com/apache/incubator-flagon-useralejs/blob/7a7b2ce73963d9dd1b0bff9419ded12e15af10ad/src/packageLogs.js#L116-L118

Not passing the event into `packageCustomLog` can lead to errors when using `userale.map` and `userale.packageCustomLog` together. In the [example](https://github.com/apache/incubator-flagon-useralejs/pull/64#discussion_r597152859) mentioned in Rob's comment above, we check the properties of the `e` argument to determine if we need to add custom labeling. However, if you setup the `userale.map` function this way and call `packageCustomLog`, then the `e` argument never gets passed to the `window.userale.map` function, resulting in an error when `e` properties are accessed as `e` does not exist.

In the latest commit to this PR I get around this issue by checking for `e` before accessing its properties, but this feels a bit hacky.

If I'm understanding everything correctly, we should probably either add `e` as an argument to the mapHandler in `packageCustomLog` or provide a way to bypass the mapHandler in that function.

**Comment by confusingstraw** [comment]
i think checking `e` isn't too egregious. it just means it is an optional argument. i do see your point though, and i am not sure where i land on the issue. i don't think we ought to add `e` to `packageCustomLog`, as that completely changes the nature of its API. i think the main question is whether the `map` handler ought to be called from within `packageCustomLog`. i'd argue yes, but i could be convinced otherwise.

you don't actually _have_ to check whether `e` is passed either. you can just see if `log.logType === 'custom'`. doesn't change the argument too much though.

**Comment by UncleGedd** [comment]
I understand. Looking through the README it seems the intention of `packageCustomLog` is to log non-event data, which makes sense. I also agree that it makes sense to put the onus on the user to validate that their custom mapHandler is compatible with `packageCustomLog`.  It does kind of seem like a "gotcha" though. It may be worth it to call out in the docs that logs packaged using `packageCustomLog`  are still subject to the Userale map/filter functions.

**Comment by confusingstraw** [comment]
i'd say the intent of `packageCustomLogs` is to implement custom or application-specific events. maybe something like "userLoggedIn" as a custom event type or similar. definitely something worth calling out in the docs.

**Pull request merged**